File: nezha/src/agents/wukong/taibai_client.py
"""
太白客户端 Python 实现
用于与太白服务（Matrix）通信
"""

import asyncio
import logging
from typing import Optional, Dict, Any

logger = logging.getLogger(__name__)


class TaibaiClient:
    """
    太白客户端 - 对接太白 SDK (Matrix 协议)
    
    用于发送消息到 Matrix 房间
    """
    
    def __init__(
        self,
        base_url: str = "http://localhost:8081",
        token: Optional[str] = None,
    ):
        """
        初始化太白客户端
        
        Args:
            base_url: 太白服务地址
            token: 认证 token
        """
        self.base_url = base_url.rstrip("/")
        self.token = token or ""
        self._session: Optional[Any] = None
        
    async def _get_session(self) -> Any:
        """获取 HTTP 会话"""
        if self._session is None:
            import aiohttp
            self._session = aiohttp.ClientSession()
        return self._session
        
    async def _request(
        self,
        method: str,
        path: str,
        body: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """
        发送 HTTP 请求
        
        Args:
            method: HTTP 方法
            path: 请求路径
            body: 请求体
            
        Returns:
            响应数据
        """
        import aiohttp
        
        session = await self._get_session()
        url = f"{self.base_url}{path}"
        
        headers = {
            "Content-Type": "application/json",
        }
        
        if self.token:
            headers["Authorization"] = f"Bearer {self.token}"
            
        try:
            async with session.request(
                method,
                url,
                json=body,
                headers=headers,
            ) as response:
                if response.status >= 400:
                    text = await response.text()
                    raise Exception(f"Taibai API error: {response.status} - {text}")
                    
                if response.content_type == "application/json":
                    return await response.json()
                return {}
                
        except aiohttp.ClientError as e:
            logger.error(f"HTTP request failed: {e}")
            raise
            
    async def ping(self) -> bool:
        """
        测试连接
        
        Returns:
            是否连接成功
        """
        try:
            await self._request("GET", "/_matrix/client/versions")
            return True
        except Exception as e:
            logger.warning(f"Ping failed: {e}")
            return False
            
    async def send_text_message(
        self,
        room_id: str,
        content: str,
    ) -> Dict[str, Any]:
        """
        发送文本消息
        
        Args:
            room_id: 房间 ID
            content: 消息内容
            
        Returns:
            响应包含 event_id
        """
        import uuid
        
        txn_id = f"m{uuid.uuid4().hex[:8}"
        body = {
            "msgtype": "m.text",
            "format": "plain",
            "body": content,
        }
        
        return await self._request(
            "POST",
            f"/_matrix/client/r0/rooms/{room_id}/send/m.room.message/{txn_id}",
            body,
        )
        
    async def send_html_message(
        self,
        room_id: str,
        content: str,
        html: str,
    ) -> Dict[str, Any]:
        """
        发送 HTML 消息
        
        Args:
            room_id: 房间 ID
            content: 纯文本内容
            html: HTML 内容
            
        Returns:
            响应包含 event_id
        """
        import uuid
        
        txn_id = f"m{uuid.uuid4().hex[:8}"
        body = {
            "msgtype": "m.text",
            "format": "org.matrix.html",
            "body": content,
            "formatted_body": html,
        }
        
        return await self._request(
            "POST",
            f"/_matrix/client/r0/rooms/{room_id}/send/m.room.message/{txn_id}",
            body,
        )
        
    async def send_card_message(
        self,
        room_id: str,
        card: Dict[str, Any],
    ) -> Dict[str, Any]:
        """
        发送卡片消息
        
        Args:
            room_id: 房间 ID
            card: 卡片内容
            
        Returns:
            响应包含 event_id
        """
        import uuid
        import json
        
        txn_id = f"m{uuid.uuid4().hex[:8}"
        body = {
            "msgtype": "m.custom.card",
            "format": "plain",
            "body": json.dumps(card),
            "content": card,
        }
        
        return await self._request(
            "POST",
            f"/_matrix/client/r0/rooms/{room_id}/send/m.room.message/{txn_id}",
            body,
        )
        
    async def send_typing(
        self,
        room_id: str,
        user_id: str,
        timeout: int = 30000,
    ) -> Dict[str, Any]:
        """
        发送正在输入状态
        
        Args:
            room_id: 房间 ID
            user_id: 用户 ID
            timeout: 超时时间（毫秒）
            
        Returns:
            响应
        """
        body = {
            "timeout": timeout,
        }
        
        return await self._request(
            "PUT",
            f"/_matrix/client/r0/rooms/{room_id}/typing/{user_id}",
            body,
        )
        
    async def close(self) -> None:
        """关闭会话"""
        if self._session:
            await self._session.close()
            self._session = None


class HTTPMessageSender:
    """
    HTTP 消息发送器（无太白客户端时的后备方案）
    通过天枢的 HTTP API 发送消息
    """
    
    def __init__(
        self,
        tianshu_url: str = "http://localhost:8082",
        access_token: Optional[str] = None,
    ):
        self.tianshu_url = tianshu_url.rstrip("/")
        self.access_token = access_token or ""
        
    async def send_message(
        self,
        room_id: str,
        content: str,
        message_type: str = "text",
    ) -> Optional[str]:
        """
        发送消息
        
        Args:
            room_id: 房间 ID
            content: 消息内容
            message_type: 消息类型
            
        Returns:
            消息 ID 或 None
        """
        import aiohttp
        
        # 尝试通过天枢 API 发送
        url = f"{self.tianshu_url}/api/v1/delivery/message"
        
        headers = {"Content-Type": "application/json"}
        if self.access_token:
            headers["Authorization"] = f"Bearer {self.access_token}"
            
        body = {
            "room_id": room_id,
            "content": content,
            "message_type": message_type,
        }
        
        try:
            async with aiohttp.ClientSession() as session:
                async with session.post(url, json=body, headers=headers) as response:
                    if response.status == 200:
                        data = await response.json()
                        return data.get("event_id")
                    else:
                        logger.warning(f"HTTP message send failed: {response.status}")
                        return None
        except Exception as e:
            logger.error(f"HTTP message send error: {e}")
            return None

File: src/clients/TaibaiClient.ts
/**
 * TaibaiClient - 太白客户端
 * 负责与太白服务通信：消息发送、房间管理
 * 对应 Go SDK: taibai/sdk/go/
 */

import { EventEmitter } from 'events';

export interface TaibaiConfig {
  serverAddress: string;  // 太白服务地址 (e.g., http://localhost:8083)
  token?: string;         // 认证 token
  apiKey?: string;        // API Key
}

export interface SendMessageRequest {
  roomId: string;
  content: string;
  messageType?: string;   // 默认: m.text
  format?: string;        // 默认: plain
  body?: string;
}

export interface SendMessageResponse {
  event_id: string;
  room_id: string;
  sender?: string;
  timestamp?: number;
}

export interface MessageEvent {
  event_id: string;
  room_id: string;
  sender: string;
  type: string;
  timestamp: number;
  content: Record<string, unknown>;
}

export interface MessagesResponse {
  chunk: MessageEvent[];
  start: string;
  end: string;
}

/**
 * Taibai 客户端 - 对接太白 SDK (Matrix 协议)
 * 
 * 太白 SDK 使用 Matrix 协议进行消息传递:
 * - POST /_matrix/client/r0/rooms/{roomId}/send/m.room.message
 * - GET /_matrix/client/r0/rooms/{roomId}/messages
 */
export class TaibaiClient extends EventEmitter {
  private config: TaibaiConfig;
  private baseUrl: string;
  private token: string;

  constructor(config: TaibaiConfig) {
    super();
    this.config = config;
    this.baseUrl = this.normalizeUrl(config.serverAddress);
    this.token = config.token || config.apiKey || '';
  }

  private normalizeUrl(url: string): string {
    // 确保 URL 有 http:// 前缀
    if (!url.startsWith('http://') && !url.startsWith('https://')) {
      url = 'http://' + url;
    }
    // 移除尾部斜杠
    return url.replace(/\/$/, '');
  }

  private async request<T>(
    method: string,
    path: string,
    body?: unknown,
    query?: Record<string, string>
  ): Promise<T> {
    let url = `${this.baseUrl}${path}`;
    
    // 添加查询参数
    if (query) {
      const params = new URLSearchParams(query);
      url += '?' + params.toString();
    }

    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };

    if (this.token) {
      headers['Authorization'] = `Bearer ${this.token}`;
    }

    const response = await fetch(url, {
      method,
      headers,
      body: body ? JSON.stringify(body) : undefined,
    });

    if (!response.ok) {
      const errorBody = await response.json().catch(() => ({})) as Record<string, unknown>;
      throw new Error(
        `Taibai API error: ${response.status} ${response.statusText} - ${(errorBody.error as string) || (errorBody.message as string) || 'Unknown error'}`
      );
    }

    // 处理空响应
    const text = await response.text();
    if (!text) {
      return {} as T;
    }

    return JSON.parse(text) as T;
  }

  /**
   * 发送文本消息到房间
   */
  async sendTextMessage(roomId: string, content: string): Promise<SendMessageResponse> {
    return this.sendMessage({
      roomId,
      content,
      messageType: 'm.text',
      format: 'plain',
    });
  }

  /**
   * 发送 HTML 消息到房间
   */
  async sendHTMLMessage(roomId: string, content: string, html: string): Promise<SendMessageResponse> {
    return this.sendMessage({
      roomId,
      content,
      body: content,  // plain text fallback
      messageType: 'm.text',
      format: 'html',
    });
  }

  /**
   * 发送卡片消息到房间 (Matrix card)
   */
  async sendCardMessage(roomId: string, card: Record<string, unknown>): Promise<SendMessageResponse> {
    return this.sendMessage({
      roomId,
      content: JSON.stringify(card),
      messageType: 'm.custom.card',  // 自定义消息类型
      format: 'plain',
    });
  }

  /**
   * 发送消息到房间
   */
  async sendMessage(req: SendMessageRequest): Promise<SendMessageResponse> {
    const messageType = req.messageType || 'm.text';
    const format = req.format || 'plain';
    
    const body = {
      msgtype: messageType,
      format: format,
      body: req.body || req.content,
      content: req.content,
    };

    return this.request<SendMessageResponse>(
      'POST',
      `/_matrix/client/r0/rooms/${encodeURIComponent(req.roomId)}/send/m.room.message`,
      body
    );
  }

  /**
   * 获取房间消息
   */
  async getRoomMessages(
    roomId: string,
    limit: number = 20,
    from?: string
  ): Promise<MessagesResponse> {
    const query: Record<string, string> = {
      limit: String(limit),
      dir: 'b',
    };
    
    if (from) {
      query['from'] = from;
    }

    return this.request<MessagesResponse>(
      'GET',
      `/_matrix/client/r0/rooms/${encodeURIComponent(roomId)}/messages`,
      undefined,
      query
    );
  }

  /**
   * 获取特定消息
   */
  async getMessage(roomId: string, eventId: string): Promise<MessageEvent> {
    return this.request<MessageEvent>(
      'GET',
      `/_matrix/client/r0/rooms/${encodeURIComponent(roomId)}/event/${encodeURIComponent(eventId)}`
    );
  }

  /**
   * 删除消息 (redact)
   */
  async redactMessage(roomId: string, eventId: string, reason?: string): Promise<void> {
    const body = reason ? { reason } : {};
    
    await this.request(
      'PUT',
      `/_matrix/client/r0/rooms/${encodeURIComponent(roomId)}/redact/${encodeURIComponent(eventId)}`,
      body
    );
  }

  /**
   * 创建房间
   */
  async createRoom(options: {
    name?: string;
    topic?: string;
    isPrivate?: boolean;
  }): Promise<{ room_id: string }> {
    const body = {
      name: options.name,
      topic: options.topic,
      visibility: options.isPrivate ? 'private' : 'public',
    };

    return this.request<{ room_id: string }>(
      'POST',
      '/_matrix/client/r0/createRoom',
      body
    );
  }

  /**
   * 加入房间
   */
  async joinRoom(roomIdOrAlias: string): Promise<{ room_id: string }> {
    return this.request<{ room_id: string }>(
      'POST',
      `/_matrix/client/r0/join/${encodeURIComponent(roomIdOrAlias)}`
    );
  }

  /**
   * 离开房间
   */
  async leaveRoom(roomId: string): Promise<void> {
    await this.request(
      'POST',
      `/_matrix/client/r0/rooms/${encodeURIComponent(roomId)}/leave`
    );
  }

  /**
   * 获取用户信息
   */
  async getProfile(userId: string): Promise<{
    displayname?: string;
    avatar_url?: string;
  }> {
    return this.request(
      'GET',
      `/_matrix/client/r0/profile/${encodeURIComponent(userId)}`
    );
  }

  /**
   * 测试连接
   */
  async ping(): Promise<boolean> {
    try {
      await this.request<{ flows: unknown[] }>('GET', '/_matrix/client/versions');
      return true;
    } catch {
      return false;
    }
  }
}

export default TaibaiClient;
